[Elvis Teo] iP

TLChicken opened the pull request:
# CS2103TaskDashboard

> "There is some _science_ but there is **art** as well." - Damith C. [(source)](https://www.youtube.com/watch?v=5jvOBbP6u2o)

CS2103TaskDashboard helps you overwhelm your mind by reminding you that you still have Weekly Tasks to be done whenever you look at it!

* Does not forget what you need to do.
* Tracks your weekly tasks.
* Convenient to use!

To set it up:
1. Download it from [here](https://github.com/TLChicken/ip/releases/tag/A-Jar).
1. Set it up.
1. Add your weekly tasks.
1. Now look at it every now and then to remind yourself of the tasks you need to do! 

Features:
- [ ] Track down your CS2103T Deadlines!
- [ ] Keeps track of your CS2103T Tasks!
- [ ] Look at your task list to remember what tasks you still have yet to finish! :stuck_out_tongue_winking_eye:

The best part is that CS2103TaskManager is **FREE**!!! (No school fees required!)

Also this app can help you with your Software Engineering skills just like CS2103T! I used Object-Oriented Programming, using concepts like inheritance and polymorphism in our `TodosTask`, `DeadlineTask` and `EventTask` which inherits from `BaseTask`!

Just take a look at this code:
```Java
public class CS2103T {
    private IndividualProject ip;
    private TeamProject tp;
    private TopicsList topicsList;
    
    public static void main(String[] args) {
        System.out.println("You still have " + ip.getTasksLeft + ", " + tp.getTasksLeft + ", " + topicsList.getUnreadTopics + " left to do for this week!!!!! :D ");
    }
}
```

Commit:
Add Gradle support

TLChicken committed:
Create line separator and print it

TLChicken committed:
Create runInputLoopMain fn

Currently unimplemented. Would be used to implement a loop of input detection.

TLChicken committed:
Add scanner for detecting input

TLChicken committed:
Enhance welcome message

Add info in welcome message to tell users what to do. (Give Duke something to do.)

TLChicken committed:
Add dukeSays fn to print messages between lineseps

TLChicken committed:
Add Duke echos whatever that is typed until "bye" is typed

TLChicken committed:
Add empty line padding

TLChicken committed:
Separated lineSeps into 3, make Duke printouts nicer

TLChicken committed:
Add "bye" command to exit Duke

TLChicken committed:
Add indentation to Duke's messages

TLChicken committed:
Create skeleton fn for adding things to the list

Also added @param tags to javadocs.

TLChicken committed:
Implement addToTDL fn

TLChicken committed:
Add Duke informs when adding to list

TLChicken committed:
Add check for the input "list"

TLChicken committed:
Add Listing out items of list.

TLChicken committed:
Remove indentation when Duke says something

TLChicken committed:
Remove extra newline when printing list

TLChicken committed:
Encapsulate the list in own container

TLChicken committed:
Modify Duke to work with new TDL container

TLChicken committed:
Add Duke checking for "done" command

TLChicken committed:
Add TDLTask encapsulation for use with the TDList

TLChicken committed:
Modify TDList to work with Task

TLChicken committed:
Add marking task as done

TLChicken committed:
Add error handling for invalid integer input

TLChicken committed:
Add task type enum

TLChicken committed:
Create DeadlineTask

TLChicken committed:
Create EventTask base

TLChicken committed:
Create ToDo Task base

TLChicken committed:
Add adding tasks now shows total task count

TLChicken committed:
Add task type checking

TLChicken committed:
Add ability to create different types of tasks

TLChicken committed:
Add printing of task type checkbox

TLChicken committed:
Add printing of task type checkbox to individual task types

TLChicken committed:
Move Task Type Checking to Duke input processor

TLChicken committed:
Move task type checker to TDLTask class

TLChicken committed:
Change tdlAdd to take in TaskType

TLChicken committed:
Add functionality to remove todo prefix when adding todo

TLChicken committed:
Done TextUiTesting Test 1

TLChicken committed:
Add DukeExceptionBase

TLChicken committed:
Abstract out command processing into own fn

TLChicken committed:
Add isExited in Duke class

TLChicken committed:
Add unknownCommandEntered fn that throws DukeException

TLChicken committed:
Add exception handling to Duke fns

TLChicken committed:
Add exception handling to Duke's list

TLChicken committed:
Modify setAsDone to throw exception if task was alr done

TLChicken committed:
Add functionality to delete task in list

TLChicken committed:
Add command "delete"

TLChicken committed:
Refactor into packages

Packages Created:
dukeScripts
dukeScripts.Tasks

TLChicken committed:
Change dukeScripts package to duke

TLChicken committed:
Add Duke Instance Logic

TLChicken committed:
Add parse fn to Command Parser

TLChicken committed:
Change Duke from static to instance

TLChicken committed:
Encapsulate DukeUi

Add buffer and release system for DukeUi.

TLChicken committed:
Refactor TDLTask and list tasks method name

TDLTask renamed to Base Task.
printOutTDL() renamed to printOutWholeList().

TLChicken committed:
Add getters for Duke Component Controllers

TLChicken committed:
Create save tasks method skeleton

TLChicken committed:
Create sample Save Files

TLChicken committed:
Add handling of XML with wrong format

TLChicken committed:
Add method to convert task letter to task enum

TLChicken committed:
Add new constructor to DeadlineTask

Enables setting isDone while creating the task.

TLChicken committed:
Add new constructor to ToDos Task

Enables setting isDone while creating the task.

TLChicken committed:
Add new constructor to EventTask

Enables setting isDone while creating the task.

TLChicken committed:
Modify Sample Save File

TLChicken committed:
Add load new Task List fn in list mgr

TLChicken committed:
Add loading save trigger when Duke starts

TLChicken committed:
Add loading save from XML functionality

TLChicken committed:
Add saving task list when list changes

TLChicken committed:
Add getter for extra info in tasks

TLChicken committed:
Add making new XML Document obj from list

TLChicken committed:
Add writing save file to disk

TLChicken committed:
Add LocalDate format for DeadlineTask

DeadlineTask can now store 'byWhen' in LocalDate format
when the input is in any of the following 3 formats:
* DD/MM/YY
* DD/MM/YYYY
* DD Month YYYY
When stored as a LocalDate, the list will print the
dates out in the format Day Month Year.

Otherwise it would just be stored as a String.

TLChicken committed:
Merge branch 'branch-Level-7'

TLChicken committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/duke/Tasks/DeadlineTask.java

TLChicken committed:
Add local save to gitignore

TLChicken committed:
Add JUnit Test Base

TLChicken committed:
Add more tests for DeadlineTask

After trying to fix Project Settings for almost 5 hours,
(All my import statements were suddenly not working
even though code is not changed)
I found out that the problem was with IntelliJ's cache
for this project. Solution was to Press:
File > Invalidate Cache

TLChicken committed:
Stop tracking listSave1

TLChicken committed:
Upgrade DeadlineTask Date Storage and Tests

TLChicken committed:
Change savefile location and create JAR

TLChicken committed:
Change gitignore to ignore local save

TLChicken committed:
Add more JavaDocs to everything

TLChicken committed:
Check code complies with coding standard

TLChicken committed:
Add find command to parser

TLChicken committed:
Add method to search for words in Tasks

TLChicken committed:
Add find command functionality

TLChicken committed:
Clarify keyword definition

TLChicken committed:
Merge branch 'branch-A-JavaDoc'

TLChicken committed:
Merge branch 'branch-Level-9'

TLChicken committed:
Merge branch 'branch-A-CodingStandard'

James-Kua commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,258 @@
+
+    }
+
+    /**
+     * Use to get the current instance of Duke that is running.
+     * Creates a new instance of Duke if there is none running.
+     * @return the Duke object that is running.
+     */
```
Should there be an empty line before @return?

James-Kua commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,258 @@
+    }
+
+    /**
+     * Getter for Duke's Command Parser.
```
Should the first word be a verb like 'gets'? I noticed the same issue in several other places too

James-Kua commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,258 @@
+        throw new DukeExceptionBase("Please enter something valid!");
+    }
+
+    private void addToTDL(String str, BaseTask.TaskType currTaskType) throws DukeExceptionBase {
```
Should it be addToTdl instead? Reference: Abbreviations and acronyms should not be uppercase when used as a part of a name.

James-Kua commented on src/main/java/duke/DukeListMgr.java:
```diff
@@ -0,0 +1,234 @@
+    /**
+     * Used to mark the corresponding task as done.
+     *
+     * @param taskNo The task number in the list.
+     *
+     * @return Duke's output from this command
+     *
+     * @throws DukeExceptionBase when an invalid task is specified or if task is already done.
```
Should these lines have no space between them?

James-Kua commented on src/main/java/duke/Tasks/DeadlineTask.java:
```diff
@@ -0,0 +1,154 @@
+            this.byWhenDate = byDate;
+
+        } catch (DateTimeParseException e) {
+            throw new DukeExceptionBase("Deadline Date does not follow the format required. " +
```
Should the operator '+' be included in the next line instead? Reference: Break before an operator.

TLChicken replied:
Thanks, I have fixed all 5 of them.

James-Kua reviewed (commented):
I think the there's some inconsistencies in the JavaDocs and it is recurring in different parts. I think it just needs some cleaning up and it should be good to go! Also spotted some minor issues as well.

TLChicken committed:
Merge 'origin/add-gradle-support' into local

TLChicken committed:
Add Gradle

TLChicken committed:
Add CheckStyle

TLChicken committed:
Add JavaFX to build.gradle

TLChicken committed:
Make BaseTask abstract

TLChicken committed:
Fix some style errors

TLChicken committed:
Move Tasks into Duke package

TLChicken committed:
Delete debug code from DeadlineTask

TLChicken committed:
Fix styling

TLChicken committed:
Add JavaFX

TLChicken committed:
Add functional GUI

TLChicken committed:
Fix Welcome Msg and Label Size

TLChicken committed:
Fix checkstyle errors

TLChicken committed:
Merge branch 'branch-Level-10'

TLChicken committed:
Change GUI font

Currently, the font used by the GUI is the default font.

It needs to change because the current font is:
* hard to align in a list as characters have different sizes.
* not able to work on some Macs.

This commit will change the font of the entire
GUI from the default font to Courier New.

This is because Courier New...:
* is monospaced, making the list neater.
* is a default font on most systems.
* fixes a font bug on some Macs.

TLChicken committed:
Add assertions

Currently there are no assertions in the code.

This needs to change because we need another safeguard
to detect bugs in the code.

Let's add assertions.

This is to verify the assumptions that are made at certain
points of the code to make sure they remain valid.

TLChicken committed:
Improve code quality

Code quality needs to be improved to make the code
more human-friendly and readable.

Let's refactor the code, extract methods to make
long methods shorter, remove dead code, and make
the happier path prominent.

TLChicken committed:
Merge pull request #1 from TLChicken/branch-A-CodeQuality

Improve code quality

TLChicken committed:
Merge branch 'master' of github.com:TLChicken/ip

TLChicken committed:
Merge branch 'master' into branch-A-Assertions

TLChicken committed:
Merge pull request #2 from TLChicken/branch-A-Assertions

Merge A-Assertions

TLChicken committed:
Add C-Help

TLChicken committed:
Improve GUI

Add background, change text color

TLChicken committed:
Add bold to dialog text label

TLChicken committed:
Amend help screen

TLChicken committed:
Fix Checkstyle errors

TLChicken committed:
Add icon and name to window

TLChicken committed:
Add Ui image to docs folder

TLChicken committed:
Add commands to User Guide

TLChicken committed:
Fix Javadocs
